python/ray/util/worker_group.py

```diff
@@ -0,0 +1,308 @@
+
+    def execute_async(self, func: Callable[..., T], *args,
+                      **kwargs) -> List[ObjectRef]:
+        """Execute ``func`` on each worker and return the futures.
+
+        Args:
+            func (Callable): A function to call on each worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (List[ObjectRef]) A list of ``ObjectRef`` representing the
+                output of ``func`` from each worker. The order is the same
+                as ``self.workers``.
+
+        """
+        if len(self.workers) <= 0:
+            raise RuntimeError("There are no active workers. This worker "
+                               "group has most likely been shut down. Please"
+                               "create a new WorkerGroup or restart this one.")
+
+        return [
+            w.actor._BaseWorkerMixin__execute.remote(func, *args, **kwargs)
+            for w in self.workers
+        ]
+
+    def execute(self, func: Callable[..., T], *args, **kwargs) -> List[T]:
+        """Execute ``func`` on each worker and return the outputs of ``func``.
+
+        Args:
+            func (Callable): A function to call on each worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (List[T]) A list containing the output of ``func`` from each
+                worker. The order is the same as ``self.workers``.
+
+        """
+        return ray.get(self.execute_async(func, *args, **kwargs))
+
+    def execute_single_async(self, worker_index: int, func: Callable[..., T],
+                             *args, **kwargs) -> ObjectRef:
+        """Execute ``func`` on worker ``worker_index`` and return futures.
+
+        Args:
+            worker_index (int): The index to execute func on.
+            func (Callable): A function to call on the first worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (ObjectRef) An ObjectRef representing the output of func.
+
+        """
+        if worker_index >= len(self.workers):
+            raise ValueError(f"The provided worker_index {worker_index} is "
+                             f"not valid for {self.num_workers} workers.")
+        return self.workers[worker_index].actor._BaseWorkerMixin__execute\
+            .remote(
+            func, *args, **kwargs)
+
+    def execute_single(self, worker_index: int, func: Callable[..., T], *args,
+                       **kwargs) -> T:
+        """Execute ``func`` on worker with index ``worker_index``.
+
+        Args:
+            worker_index (int): The index to execute func on.
+            func (Callable): A function to call on the first worker.
+            args, kwargs: Passed directly into func.
+
+        Returns:
+            (T) The output of func.
+
+        """
+
+        return ray.get(
+            self.execute_single_async(worker_index, func, *args, **kwargs))
+
+    def remove_workers(self, worker_indexes: List[int]):
```

Comment:
```suggestion
    def remove_workers(self, worker_indices: List[int]):
```

Reply:
Should worker_indices be optional?

Reply:
Just for clarification, what would be the behavior if it's not provided. Would it be all workers are removed, or just the last worker?

Reply:
Ah, it doesn't make much sense as-written but I was imagining something like "remove N workers, don't care which." Let's just leave it as-is for now.